ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/LMDB.hs

```diff
@@ -0,0 +1,372 @@
+
+
+lmdbCopy :: Db l -> FS.SomeHasFS m -> HD.BackingStorePath -> m ()
+lmdbCopy = error "unimplemented"
+
+-- | How to initialise an LMDB Backing store
+data LMDBInit l
+  = LIInitialiseFromMemory (WithOrigin SlotNo) (Tables.LedgerTables l Tables.ValuesMK) -- ^ Initialise with these values
+  | LIInitialiseFromLMDB FS.FsPath -- ^ Initialise by copying from an LMDB db at this path
+  | LINoInitialise -- ^ The database is already initialised
+
+-- | Initialise a backing store
+newLMDBBackingStore :: forall l. (Tables.TableStuff l)
+  => FS.SomeHasFS IO 
+  -> FS.FsPath -- ^ The path for the store we are opening
+  -> LMDBInit l 
+  -> IO (LMDBBackingStore l)
+newLMDBBackingStore sfs path init_db = do
+  let
+    limits = defaultLMDBLimits
+    (gdd, copy_db_action, init_action) = case init_db of
+      LIInitialiseFromLMDB fp
+        -- If fp == path then this is the LINoInitialise case
+        | fp /= path -> (GDDMustNotExist, initFromLMDBs sfs fp path, \_ -> pure ())
+      LIInitialiseFromMemory slot vals -> (GDDMustNotExist, pure (), initFromVals slot vals)
+      _ -> (GDDMustExist, pure (), \_ -> pure ())
+  -- get the filepath for this db creates the directory if appropriate
+  dbFilePath <- guardDbDir gdd sfs path
+
+  -- copy from another lmdb path if appropriate
+  copy_db_action
+
+  -- open this database
+  dbEnv <- LMDB.openEnvironment dbFilePath limits
+
+  -- the LMDB.Database that holds the DbState (i.e. sequence number)
+  dbstate_db <- LMDB.readOnlyTransaction dbEnv $ LMDB.getDatabase (Just "_dbstate")
+  dbSettings <- LMDB.newDBRef dbEnv dbstate_db (0 :: Int)
+
+  dbValueHandles <- IOLike.newTMVarIO Map.empty
+
+  -- Here we get the LMDB.Databases for the tables of the ledger state
+  dbBackingTables <- LMDB.readOnlyTransaction dbEnv $
+    Tables.traverseLedgerTables getDb_amk Tables.namesLedgerTables
+
+  let
+    db = Db{..}
+    bsClose = do
+      mb_vhs <- IOLike.atomically $ IOLike.tryTakeTMVar dbValueHandles
+      for_ mb_vhs $ traverse forceCloseValueHandle
+
+    bsCopy = lmdbCopy db
+    
+    bsValueHandle = getLMDBBackingStoreValueHandle db
+
+    bsWrite :: SlotNo -> Tables.LedgerTables l Tables.DiffMK -> IO ()
+    bsWrite slot diffs = LMDB.readWriteTransaction dbEnv $
+      withDbSettingsRW dbSettings $ \s@DbState{dbsSeq} -> do
+        when (At slot <= dbsSeq) $ Exn.throw $ DbErrNonMonotonicSeq (At slot) dbsSeq
+        void $ Tables.zip2ALedgerTables writeLMDBTable_amk dbBackingTables diffs
+        pure ((), s {dbsSeq = At slot})
+
+  -- now initialise those tables if appropriate
+  init_action db
+
+  pure HD.BackingStore{..}
+
+
+-- | ValueHandles hold an Async which is holding a transaction open.
+-- We maintain a map of them in the Db, (keyed on their Seq No) and
+-- they have a reference counting TVar.
+-- This means that multiple BackingStoreValueHandles will reference the
+-- same ValueHandle and thus we keep only one transaction per seq no alive.
+data ValueHandle = ValueHandle
```

Comment:
Is there a limit on how many read txs can be open? I don't expect there will be many value handles open at a time. All the ones I've used so far should be ephemeral, except for the `LocalStateQuery` mini protocol server, but there should only ever be one of those, and it's a trusted client, so we can expect they won't hold the vh open too long.

So maybe the sharing and refcounting is unnecessary complexity?

(I can imagine other reasons it's necessary; just hoping for simpler 😄.)

Reply:
If it's just because `lmdb-simple` doesn't export its internals, then maintaining a simple fork that just exposes its internal modules instead of jumping through this kind of `Async` hoop might be worth a while soon enough.

Reply:
(I'm leaving this Conversation open as a TODO.)